src/System.Net.Primitives/src/System/Net/CookieCollection.cs

```diff
@@ -0,0 +1,297 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Collections;
+using System.Globalization;
+
+namespace System.Net
+{
+    // CookieCollection
+    //
+    // A list of cookies maintained in Sorted order. Only one cookie with matching Name/Domain/Path 
+    public class CookieCollection : ICollection
+    {
+        internal enum Stamp
+        {
+            Check = 0,
+            Set = 1,
+            SetToUnused = 2,
+            SetToMaxUsed = 3,
+        }
+
+        internal int _version;
+        private ArrayList _list = new ArrayList();
+
+        private DateTime _timeStamp = DateTime.MinValue;
+        private bool _hasOtherVersions;
+
+        private bool _isReadOnly;
+
+        public CookieCollection()
+        {
+            _isReadOnly = true;
+        }
+
+        internal CookieCollection(bool IsReadOnly)
+        {
+            _isReadOnly = IsReadOnly;
+        }
+
+        public bool IsReadOnly
+        {
+            get
+            {
+                return _isReadOnly;
+            }
+        }
+
+        public Cookie this[int index]
+        {
+            get
+            {
+                if (index < 0 || index >= _list.Count)
+                {
+                    throw new ArgumentOutOfRangeException("index");
+                }
+                return (Cookie)(_list[index]);
+            }
+        }
+
+        public Cookie this[string name]
+        {
+            get
+            {
+                foreach (Cookie c in _list)
+                {
+                    if (string.Compare(c.Name, name, StringComparison.OrdinalIgnoreCase) == 0)
+                    {
+                        return c;
+                    }
+                }
+                return null;
+            }
+        }
+
+        public void Add(Cookie cookie)
+        {
+            if (cookie == null)
+            {
+                throw new ArgumentNullException("cookie");
+            }
+            ++_version;
+            int idx = IndexOf(cookie);
+            if (idx == -1)
+            {
+                _list.Add(cookie);
+            }
+            else
+            {
+                _list[idx] = cookie;
+            }
+        }
+
+        public void Add(CookieCollection cookies)
```

Comment:
How common is it to have large numbers of cookies?  As implemented, this is an ```O(N^2)``` operation in the number of cookies.